cli/cmd/diagnostics.go

```diff
@@ -0,0 +1,205 @@
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	adminHTTPPortName string = "admin-http"
+)
+
+type diagnosticsResult struct {
+	pod       string
+	container string
+	metrics   []byte
+	err       error
+}
+type diagResult []diagnosticsResult
+
+func (s diagResult) Len() int {
+	return len(s)
+}
+func (s diagResult) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+func (s diagResult) Less(i, j int) bool {
+	if s[i].pod != s[j].pod {
+		return s[i].pod < s[j].pod
+	}
+	return s[i].container < s[j].container
+}
+
+type diagnosticsOptions struct {
+	wait time.Duration
+}
+
+func newDiagnosticsOptions() *diagnosticsOptions {
+	return &diagnosticsOptions{
+		wait: 300 * time.Second,
+	}
+}
+
+func newCmdDiagnostics() *cobra.Command {
+	options := newDiagnosticsOptions()
+
+	cmd := &cobra.Command{
+		Use:   "diagnostics",
+		Short: "Fetch metrics directly from the Linkerd control plane containers",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			k8sAPI, err := k8s.NewAPI(kubeconfigPath, kubeContext, impersonate, 0)
+			if err != nil {
+				return err
+			}
+
+			timeoutSeconds := int64(30)
+			deployments, err := k8sAPI.AppsV1().Deployments(controlPlaneNamespace).List(metav1.ListOptions{TimeoutSeconds: &timeoutSeconds})
+			if err != nil {
+				//TODO: handle error
+				return err
+			}
+
+			// ensure we can connect to the public API before fetching the diagnostics.
+			checkPublicAPIClientOrRetryOrExit(time.Now().Add(options.wait), true)
+
+			resCount := 0
+			resultChan := make(chan diagnosticsResult)
+			for _, d := range deployments.Items {
+				pods, err := getPodsFor(k8sAPI, controlPlaneNamespace, d.Name)
+				if err != nil {
+					//TODO: handle error
+				}
+
+				for _, pod := range pods {
+					containers, err := getNonProxyContainersForPod(pod)
+					if err != nil {
+						// TODO: handle error
+					}
+
+					for i := range containers {
+						resCount++
+						go func(container corev1.Container) {
+							bytes, err := getDiagnostics(k8sAPI, pod, container, verbose)
+
+							resultChan <- diagnosticsResult{
+								pod:       pod.GetName(),
+								container: container.Name,
+								metrics:   bytes,
+								err:       err,
+							}
+						}(containers[i])
+					}
+				}
+			}
+
+			var results []diagnosticsResult
+			timer := time.NewTimer(50 * time.Second)
+			timedOut := false
+
+			for {
+				select {
+				case result := <-resultChan:
+					results = append(results, result)
+				case <-timer.C:
+					timedOut = true
+				}
+				if len(results) == resCount || timedOut {
```

Comment:
`resCount` stores the count of the total number of containers. Any suggestions on better ways to do this?

Reply:
I think that [waitgroups](https://gobyexample.com/waitgroups) are often used for this kind of thing.